File: sdks/nodejs/nodes/object-merge/object-merge.process.js
export default async ({ inputs, settings, config }) => {
  const objects = inputs.objects;
  const arrayMode = inputs.array_mode ?? "replace";

  // Normalize to a flat array. A single connection can deliver an array of
  // objects, and a multi-connection input can deliver arrays alongside plain
  // objects, so flatten one level before filtering.
  const objectList = (Array.isArray(objects) ? objects : [objects]).flatMap(
    (entry) => (Array.isArray(entry) ? entry : [entry])
  );

  // Filter out non-objects
  const validObjects = objectList.filter(
    (obj) => obj !== null && typeof obj === "object" && !Array.isArray(obj)
  );

  if (validObjects.length === 0) {
    return { merged: {} };
  }

  const isPlainObject = (val) =>
    val !== null && typeof val === "object" && !Array.isArray(val);

  const deepMerge = (target, source) => {
    const result = { ...target };

    for (const key of Object.keys(source)) {
      const targetVal = target[key];
      const sourceVal = source[key];

      if (Array.isArray(sourceVal)) {
        if (arrayMode === "concat" && Array.isArray(targetVal)) {
          result[key] = [...targetVal, ...sourceVal];
        } else {
          result[key] = [...sourceVal];
        }
      } else if (isPlainObject(sourceVal)) {
        if (isPlainObject(targetVal)) {
          result[key] = deepMerge(targetVal, sourceVal);
        } else {
          result[key] = deepMerge({}, sourceVal);
        }
      } else {
        result[key] = sourceVal;
      }
    }

    return result;
  };

  let merged = {};
  for (const obj of validObjects) {
    merged = deepMerge(merged, obj);
  }

  return { merged };
};

File: nodes/object-merge/object-merge.process.py
from typing import Any


async def process(
    *,
    inputs: dict[str, Any],
    settings: dict[str, Any],
    config: dict[str, Any],
    node_config: dict[str, Any],
) -> dict[str, Any]:
    objects = inputs.get("objects")
    array_mode = inputs.get("array_mode") if inputs.get("array_mode") is not None else "replace"

    # Normalize to a flat list. A single connection can deliver a list of
    # objects, and a multi-connection input can deliver lists alongside plain
    # objects, so flatten one level before filtering.
    raw_list = objects if isinstance(objects, list) else [objects]
    object_list = []
    for entry in raw_list:
        if isinstance(entry, list):
            object_list.extend(entry)
        else:
            object_list.append(entry)

    # Filter out non-dicts
    valid_objects = [
        obj for obj in object_list
        if isinstance(obj, dict)
    ]

    if len(valid_objects) == 0:
        return {"merged": {}}

    def is_plain_object(val):
        return isinstance(val, dict)

    def deep_merge(target, source):
        result = {**target}

        for key in source:
            target_val = target.get(key)
            source_val = source[key]

            if isinstance(source_val, list):
                if array_mode == "concat" and isinstance(target_val, list):
                    result[key] = target_val + source_val
                else:
                    result[key] = source_val.copy()
            elif is_plain_object(source_val):
                if is_plain_object(target_val):
                    result[key] = deep_merge(target_val, source_val)
                else:
                    result[key] = deep_merge({}, source_val)
            else:
                result[key] = source_val

        return result

    merged = {}
    for obj in valid_objects:
        merged = deep_merge(merged, obj)

    return {"merged": merged}
